### PR TITLE
[FLINK-17579] Allow user to set the prefix of TaskManager's ResourceID in standalone mode

### PR DESCRIPTION
**Pull request by KarmaGYZ**
## What is the purpose of the change

Set the resource id of taskexecutor according to environment variable if exist in standalone mode.

## Brief change log

Allow user to set the resource id of taskexecutor according to environment variable if exist in standalone mode.
The relevant env variable is "FLINK_STANDALONE_TASK_EXECUTOR_ID".

## Verifying this change

It could be verified manually.
- `bin/jobmanager.sh start`
- `FLINK_STANDALONE_TASK_EXECUTOR_ID=taskManager-1 bin/taskmanager.sh`

`2020-05-09 18:00:09,856 INFO  org.apache.flink.runtime.taskexecutor.TaskManagerRunner      [] - Starting TaskManager with ResourceID: taskManager-1` occurred in the log file of task executor.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 1]
```diff
@@ -293,7 +295,15 @@ public static void main(String[] args) throws Exception {
 			LOG.info("Cannot determine the maximum number of open file descriptors");
 		}
 
-		runTaskManagerSecurely(args, ResourceID.generate());
+		final Configuration config = loadConfiguration(args);
+		final String resourceIDPrefix = config.get(TaskManagerOptions.TASK_MANAGER_RESOURCE_ID_PREFIX);
+		ResourceID taskManagerResourceID;
```
Comment (wangyang0918) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:300]: Could we wrap the logic to get the `resourceID` into a separate method? Also we could add some tests for it.
* Do not specify prefix
* Specify prefix
  * get more than once and should have two different resource id

File: flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java [hunk 2]
```diff
@@ -257,6 +257,18 @@
...
+	/**
+	 * The config options defined the prefix of the TaskManager's ResourceID. If configured, the ResourceID will be spliced
+	 * with the prefix and an uuid. Otherwise, the ResourceID would be a random hex string. Only valid in standalone mode.
+	 */
+	@Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER)
+	public static final ConfigOption<String> TASK_MANAGER_RESOURCE_ID_PREFIX =
+		key("taskmanager.resource-id.prefix")
+			.stringType()
+			.noDefaultValue()
+			.withDescription("The config options defined the prefix of the TaskManager's ResourceID. If configured, the ResourceID "
```
Comment (wangyang0918) [inline flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:269]: "The prefix of the TaskManager's ResourceID. If configured, the ResourceID will be generated with the prefix and a 6-character random string. Otherwise, a 32-character random string will be used. Only valid in standalone mode."

**Comment by KarmaGYZ** [comment]
Thanks for the review! @wangyang0918 I've addressed your comments.

File: flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java [hunk 3]
```diff
@@ -257,6 +257,18 @@
 					text("\"ip\" - uses host's ip address as binding address"))
 				.build());
 
+	/**
```
Comment (wangyang0918) [inline flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:260]: nit: Update the comments or remove them.

**Comment by KarmaGYZ** [comment]
cc @tillrohrmann Would you like to take a look?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 4]
```diff
@@ -413,9 +413,10 @@ static BackPressureSampleService createBackPressureSampleService(
 	 * @param configuration The configuration for the TaskManager.
 	 * @param haServices to use for the task manager hostname retrieval
 	 */
-	public static RpcService createRpcService(
-			final Configuration configuration,
-			final HighAvailabilityServices haServices) throws Exception {
+	@VisibleForTesting
+	static RpcService createRpcService(
+		final Configuration configuration,
+		final HighAvailabilityServices haServices) throws Exception {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:419]: Why is this change necessary?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 5]
```diff
@@ -306,10 +308,8 @@ public static void runTaskManager(Configuration configuration, ResourceID resour
 		taskManagerRunner.start();
 	}
 
-	public static void runTaskManagerSecurely(String[] args, ResourceID resourceID) {
+	public static void runTaskManagerSecurely(Configuration configuration, ResourceID resourceID) {
 		try {
-			final Configuration configuration = loadConfiguration(args);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:311]: I would put the `ResourceID` generation here instead of moving the configuration loading out of this method.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 6]
```diff
@@ -91,7 +93,8 @@
 
 	private static final int STARTUP_FAILURE_RETURN_CODE = 1;
 
-	public static final int RUNTIME_FAILURE_RETURN_CODE = 2;
+	@VisibleForTesting
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:96]: Why is this necessary?

File: flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java [hunk 7]
```diff
@@ -257,6 +257,18 @@
...
+	 * The prefix of the TaskManager's ResourceID. If configured, the ResourceID will be generated with the prefix and a
+	 * 6-character random string. Otherwise, a random string will be used. Only valid in standalone mode.
+	 */
+	@Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER)
+	public static final ConfigOption<String> TASK_MANAGER_RESOURCE_ID_PREFIX =
+		key("taskmanager.resource-id.prefix")
+			.stringType()
+			.noDefaultValue()
+			.withDescription("The prefix of the TaskManager's ResourceID. If configured, the ResourceID will be generated with the prefix "
+				+ "and a 6-character random string. Otherwise, a random string will be used. Only valid in standalone mode.");
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:270]: Why is this only valid in standalone mode? Shouldn't we have a general mean how to configure the resource id for `TaskExecutors` in general? I believe it should work across the different integrations (Yarn, Mesos, K8s, standalone) all the same. The only difference would be to what value the different implementations set it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 6 above)
Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:96]: I saw this field only used in `TaskManagerRunner.java` and testing code. So I think it probably not makes sense to set the access modifier to `public`. However, we could not set it to `private` as it is used by testing code. So, I change it to `package-private` and annotate it with `VisibleForTesting`. It just a hotfix though.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 4 above)
Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:419]: Same as above. I just think it would be good to limit accessibility.

File: flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java (hunk 7 above)
Comment (wangyang0918) [reply flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:270]: Thanks for raising this valid concern. For standalone cluster, we could set different `TASK_MANAGER_RESOURCE_ID_PREFIX` per `TaskExecutor`. Currently, for active resourcemanager(Yarn, Mesos, K8s), we will generate a meaningful `ResourceID`(e.g. container id in Yarn, pod name in native K8s) per `TaskExecutor`. If we also want to respect the `TASK_MANAGER_RESOURCE_ID_PREFIX` config option, then it will be difficult to use the `ResourceID` to find the corresponding container/pod.

Given that `TASK_MANAGER_RESOURCE_ID_PREFIX` is configured to "my-yarn-cluster-tm", then all the `TaskExecutor` share the same schema "my-yarn-cluster-tm-xxxxxx". A corresponding container id could be "container_e01_1592277741400_0043_01_000002". It is really confusing.

If you just mean to override `TASK_MANAGER_RESOURCE_ID_PREFIX` option for active deployment with current meaningful name, i think it makes sense. We just use the config option to replace the environment to pass the `ResourceID` to `TaskExecutor`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 6 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:96]: Ah ok. This makes sense. I did not see the hotfix commit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 4 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:419]: Sounds good.

File: flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java (hunk 7 above)
Comment (tillrohrmann) [reply flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:270]: Yes what I mean is to have a single mean how the `ResourceID` is configured for across the different integrations (Yarn, Mesos, K8s). Maybe we should take a step back and not start with a prefix config option but simply with `TASK_MANAGER_RESOURCE_ID` which allows to set the `ResourceID`. That way it would be simple for the RM implementations to configure the `ResourceID`. Per default it is empty and then we could take the hostname and a random value, for example.

Comment (KarmaGYZ) [reply flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:270]: Thanks for the discussion. Add environment variable `TASK_MANAGER_RESOURCE_ID` 
 also sounds good. However, as @azagrebin said in FLINK-17579
> there are plans in the community to introduce a unified approach to set Flink options via environment variables.

Not sure is it ok to apply this approach. Should we move the discussion back to FLINK-17579?

Comment (wangyang0918) [reply flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:270]: Yes, actually in FLINK-17579 we also want to start with environment instead of config option. And we could unify all the current environment name(e.g. `_FLINK_CONTAINER_ID`, `_FLINK_POD_NAME`) to `TASK_MANAGER_RESOURCE_ID`. For active resourcemanager, we generate with meaningful names and do not change the current behavior. And for standalone, it could be specified with an arbitrary string(e.g. hostname, podname, etc.).

Comment (tillrohrmann) [reply flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:270]: Why do you want to do it with an environment variable? Can't it be a configuration value? That way the user could also configure it simply in his `flink-conf.yaml`.

This would be a configuration value which is specific to the started process and which should not distributed to other processes, though. I think we already have similar cases in Flink. See `DeploymentOptionsInternal.CONF_DIR`, for example.

Comment (KarmaGYZ) [reply flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:270]: Ahh, I think I misunderstand what you mean by "step back". You mean we could set the `ResourceID` to `host-uuid` by default. But for the RM implementations, they could override it to `containerID` through a configuration option `TASK_MANAGER_RESOURCE_ID `. Do I understand it correctly?

If that is the case, user should not set `TASK_MANAGER_RESOURCE_ID` in active mode (active RMs will override this option anyway). And in standalone mode, it is user's responsibility to ensure that these ids do not conflict.

Comment (tillrohrmann) [reply flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java]: Yes, this was my idea. The down side would be that people could shoot themselves in the foot with it. But I believe that this should be fine given that it is an expert feature.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 8]
```diff
@@ -293,30 +294,32 @@ public static void main(String[] args) throws Exception {
...
 	}
 
-	public static void runTaskManagerSecurely(String[] args, ResourceID resourceID) {
+	public static void runTaskManagerSecurely(String[] args, String resourceID) {
 		try {
 			final Configuration configuration = loadConfiguration(args);
-
+			if (resourceID != null) {
+				configuration.set(TaskManagerOptions.TASK_MANAGER_RESOURCE_ID, resourceID);
+			}
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:315]: I think it would be good if setting the `TASK_MANAGER_RESOURCE_ID` happens on the `ResourceManager` side.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 9]
```diff
@@ -458,4 +462,11 @@ private static String determineTaskManagerBindAddressByConnectingToResourceManag
 		HostBindPolicy bindPolicy = HostBindPolicy.fromString(configuration.getString(TaskManagerOptions.HOST_BIND_POLICY));
 		return bindPolicy == HostBindPolicy.IP ? taskManagerAddress.getHostAddress() : taskManagerAddress.getHostName();
 	}
+
+	@VisibleForTesting
+	static ResourceID getTaskManagerResourceID(Configuration config) throws Exception {
+		final String resourceID = config.get(TaskManagerOptions.TASK_MANAGER_RESOURCE_ID);
+		return resourceID != null ?
+			new ResourceID(resourceID) : new ResourceID(InetAddress.getLocalHost().getHostName() + "-" + new AbstractID().toString().substring(0, 6));
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:470]: I'm wondering whether the address the `RpcService` binds to would be more meaningful. Moreover, we could also add the port of the `RpcService`. But as a safety net we should still add a random suffix to filter out if a TM gets restarted on the same machine with the same port.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java [hunk 10]
```diff
@@ -86,6 +89,26 @@ public void testShouldShutdownIfRegistrationWithJobManagerFails() throws Excepti
 		assertThat(statusCode, is(equalTo(TaskManagerRunner.RUNTIME_FAILURE_RETURN_CODE)));
 	}
 
+	@Test
+	public void testGenerateTaskManagerResourceIDWithPrefixSpecified() throws Exception {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java:93]: Method name seems out of date.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java [hunk 11]
```diff
@@ -86,6 +89,26 @@ public void testShouldShutdownIfRegistrationWithJobManagerFails() throws Excepti
...
+		final Configuration configuration = createConfiguration();
+		final String resourceID = "test";
+		configuration.set(TaskManagerOptions.TASK_MANAGER_RESOURCE_ID, resourceID);
+		final ResourceID taskManagerResourceID = TaskManagerRunner.getTaskManagerResourceID(configuration);
+
+		assertThat(taskManagerResourceID.toString(), equalTo(resourceID));
+	}
+
+	@Test
+	public void testGenerateTaskManagerResourceIDWithoutPrefixSpecified() throws Exception {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java:103]: Same here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 9 above)
Comment (wangyang0918) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:470]: How to set the `resourceID` when we use a random port for TaskManager rpc? Also the TaskManager could bind to an ip address. So i prefer a random uuid just as before when `TASK_MANAGER_RESOURCE_ID` is not configured.

Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:470]: By creating the `ResourceID` after the `RpcService` has been started, one could obtain this information.

Comment (wangyang0918) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:470]: Well my only concern is whether it is safe to put some special characters(e.g. `:`) in the `ResourceID`. I am not aware of this could happen.

**Comment by KarmaGYZ** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 9 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: I think this should be fine. The {{ResourceID}} is defined to contain a string field.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java [hunk 12]
```diff
@@ -86,6 +88,40 @@ public void testShouldShutdownIfRegistrationWithJobManagerFails() throws Excepti
...
+
+	@Test
+	public void testGenerateTaskManagerResourceIDWithRemoteRpcService() throws Exception {
+		final Configuration configuration = createConfiguration();
+		final String rpcAddress = "flink";
+		final int rpcPort = 9090;
+		final String taskManagerResourceID = TaskManagerRunner.getTaskManagerResourceID(configuration, rpcAddress, rpcPort);
+
+		assertThat(taskManagerResourceID, notNullValue());
+		assertThat(taskManagerResourceID.substring(0, taskManagerResourceID.length() - 7),
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java:109]: maybe use `containsString` matcher

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java [hunk 13]
```diff
@@ -86,6 +88,40 @@ public void testShouldShutdownIfRegistrationWithJobManagerFails() throws Excepti
...
+
+	@Test
+	public void testGenerateTaskManagerResourceIDWithLocalRpcService() throws Exception {
+		final Configuration configuration = createConfiguration();
+		final String rpcAddress = "";
+		final int rpcPort = -1;
+		final String taskManagerResourceID = TaskManagerRunner.getTaskManagerResourceID(configuration, rpcAddress, rpcPort);
+
+		assertThat(taskManagerResourceID, notNullValue());
+		assertThat(taskManagerResourceID.substring(0, taskManagerResourceID.length() - 7),
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java:121]: Same here with the `containsString` matcher

**Comment by KarmaGYZ** [comment]
Thanks for the review @tillrohrmann . PR updated and is rebased on the latest master.

**Comment by tillrohrmann** [comment]
Thanks @KarmaGYZ. Merging this PR now.
